src/main/scala/ch/epfl/alcmp/net/Serde.scala

```diff
@@ -23,39 +24,90 @@ object Serde {
-    override def deserialize(data: String): DoneMessage = ???
+    override def serialize(obj: DoneMessage): String = Serdes.serialize[Int](obj.id)
+    override def deserialize(data: String): DoneMessage = DoneMessage(Serdes.deserialize[Int](data))
+
+  def listOf[T](using Serde[T])(sep: String): Serde[List[T]] = new Serde[List[T]]() {
+    override def serialize(arg: List[T]): String = arg.map(elem => Serdes.serialize[T](elem)).mkString(sep)
+    override def deserialize(data: String): List[T] = data.split(sep).toList.map(elem => Serdes.deserialize[T](elem))
+  }
 
   given (TypeId => Serde[DivideMessage]) with
     override def apply(t: TypeId): Serde[DivideMessage] = new Serde[DivideMessage]:
-      override def serialize(obj: DivideMessage): String = t match
-        case TypeId.ListType => ???
-        case TypeId.MatrixType => ???
-        case TypeId.HeapType => ???
-        case TypeId.BinaryTreeType => ???
-      override def deserialize(data: String): DivideMessage = t match
-        case TypeId.ListType => ???
-        case TypeId.MatrixType => ???
-        case TypeId.HeapType => ???
+      override def serialize(obj: DivideMessage): String =
+        val b1 = StringJoiner("/")
+                .add(obj.id.toString)
+                .add(obj.depth.toString)
+                .add(obj.index.toString)
+                .add(listOf[Int](",").serialize(obj.highlights))
+
+        val b2 = StringJoiner(";")
+        for ( output <- obj.outputs) {
+          output match
+            case IList(list) => b2.add(listOf[Int](",").serialize(list))
+            case IMatrix(rows) =>
+              val b3 = StringJoiner(",,")
+              rows.foreach(list => b3.add(listOf[Int](",").serialize(list)))
+              b2.add(b3.toString)
+            case IHeap(list) => b2.add(listOf[Int](",").serialize(list))
+            case IBinaryTree(root) => ???
+        }
+        b1.add(b2.toString)
+        b1.toString
+
+      override def deserialize(data: String): DivideMessage =
+        val arr = data.split("/")
+        val (id, depth, index) = (Serdes.deserialize[Int](arr(0)), Serdes.deserialize[Int](arr(1)), Serdes.deserialize[Int](arr(2)))
+        val highlights = listOf[Int](",").deserialize(arr(3))
+        t match
+        case TypeId.ListType =>
+          val outputs = arr(4).split(";").toList.map(list => IList(listOf[Int](",").deserialize(list)))
+          DivideMessage(id, depth, index, outputs, highlights)
+        case TypeId.MatrixType =>
+          val outputs = arr(4).split(";").map(lists =>
+              IMatrix(lists.split(",,").map(list => listOf[Int](",").deserialize(list).toList).toList)
+            ).toList
+          DivideMessage(id, depth, index, outputs, highlights)
+        case TypeId.HeapType =>
+          val outputs = arr(4).split(";").toList.map(list => IHeap(listOf[Int](",").deserialize(list)))
+          DivideMessage(id, depth, index, outputs, highlights)
```

Comment:
I think it'd be worth it to have a `Serde[IList]`, `Serde[IMatrix]`, etc and just use `listOf[IList]`, `listOf[IMatrix]`, etc